src/osp/Active/SysDebugRender.cpp

```diff
@@ -42,55 +58,176 @@ using namespace Magnum::Math::Literals;
+
+    // Generate fullscreen tri for texture rendering
+    using namespace Magnum;
+    if (resources.get<GL::Mesh>("fullscreen_tri").empty())
+    {
+        Vector2 screenSize = Vector2{GL::defaultFramebuffer.viewport().size()};
+
+        float aspectRatio = screenSize.x() / screenSize.y();
+
+        std::array<float, 12> surfData
+        {
+            // Vert position    // UV coordinate
+            -1.0f,  1.0f,       0.0f,  1.0f,
+            -1.0f, -3.0f,       0.0f, -1.0f,
+             3.0f,  1.0f,       2.0f,  1.0f
+        };
+
+        GL::Buffer surface(std::move(surfData), GL::BufferUsage::StaticDraw);
+        GL::Mesh surfaceMesh;
+        surfaceMesh
+            .setPrimitive(Magnum::MeshPrimitive::Triangles)
+            .setCount(3)
+            .addVertexBuffer(std::move(surface), 0,
+                RenderTexture::Position{}, RenderTexture::TextureCoordinates{});
+        resources.add<GL::Mesh>("fullscreen_tri", std::move(surfaceMesh));
+    }
 }
 
+void SysDebugRender::configure_render_passes(ActiveScene& rScene)
+{
+    /* Configure pipeline resources */
+    
+    // Add an offscreen framebuffer
+
+    auto& resources = rScene.get_context_resources();
+
+    Vector2i viewSize = GL::defaultFramebuffer.viewport().size();
+
+    GL::Texture2D color;
+    color.setStorage(1, GL::TextureFormat::RGB8, viewSize);
+    DependRes<GL::Texture2D> colorRes =
+        resources.add<GL::Texture2D>("offscreen_fbo_color", std::move(color));
+
+    GL::Renderbuffer depthStencil;
+    depthStencil.setStorage(GL::RenderbufferFormat::Depth24Stencil8, viewSize);
+    DependRes<GL::Renderbuffer> depthStencilRes =
+        resources.add<GL::Renderbuffer>("offscreen_fbo_depthStencil", std::move(depthStencil));
+
+    GL::Framebuffer fbo(Range2Di{{0, 0}, viewSize});
+    fbo.attachTexture(GL::Framebuffer::ColorAttachment{0}, *colorRes, 0);
+    fbo.attachRenderbuffer(GL::Framebuffer::BufferAttachment::DepthStencil, *depthStencilRes);
+
+    resources.add<GL::Framebuffer>("offscreen_fbo", std::move(fbo));
+
+    /* Define render passes */
+
+    // Opaque pass
+    rScene.debug_render_add(rScene.get_render_order(),
+        "opaque_pass", "", "transparent_pass",
+        [](ActiveScene& rScene, ACompCamera& camera)
+        {
+            using namespace Magnum;
+            using Magnum::GL::Renderer;
+
+            auto& reg = rScene.get_registry();
+
+            camera.m_renderTarget->bind();
+
+            camera.m_renderTarget->clear(
+                GL::FramebufferClear::Color
+                | GL::FramebufferClear::Depth
+                | GL::FramebufferClear::Stencil);
+
+            Renderer::enable(Renderer::Feature::DepthTest);
+            Renderer::enable(Renderer::Feature::FaceCulling);
+            Renderer::disable(Renderer::Feature::Blending);
+
+            // Fetch opaque objects
+            auto opaqueView = reg.view<CompDrawableDebug, ACompTransform>(
+                entt::exclude<CompTransparentDebug>);
```

Comment:
It would probably be better to have an ACompOpaque, so that this view wouldn't need to know about the transparent component, or other cases that may pop up.

Reply:
Yes the rendering components do need some work, I haven't settled completely on how to structure them. As I've said in the past, rendering is a very difficult thing to generalize so on some level it's going to be difficult to prevent the rendering system from coupling to the others more than would usually be acceptable.